[pixels-executor] error when aggregating l_extendedprice in tpch.lineitem

Query:
```sql
select sum(l_extendedprice) as revenue from lineitem where l_shipdate >= date '1994-01-01' and l_shipdate < date '1995-01-01' group by l_shipdate;
```

Error:
```bash
2024-05-02 16:09:15,034 [] [io.pixelsdb.pixels.worker.lambda.AggregationWorker]-[ERROR] error occurred in thread: Thread-3
io.pixelsdb.pixels.worker.common.WorkerException: error during scan
	at io.pixelsdb.pixels.worker.common.BaseAggregationWorker.lambda$process$0(BaseAggregationWorker.java:178) ~[task/:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) ~[?:?]
	at java.lang.Thread.run(Unknown Source) [?:?]
Caused by: io.pixelsdb.pixels.worker.common.WorkerException: failed to read the input partial aggregation file '/pixels-turbo-intermediate/intermediate/9/aggregate_1aa1395ec8e24a8aa6593987ebc64ee6/aggregate_lineitem/0/partial_aggr' and perform aggregation
	at io.pixelsdb.pixels.worker.common.BaseAggregationWorker.aggregate(BaseAggregationWorker.java:336) ~[task/:?]
	at io.pixelsdb.pixels.worker.common.BaseAggregationWorker.lambda$process$0(BaseAggregationWorker.java:173) ~[task/:?]
	... 3 more
Caused by: java.lang.ClassCastException: class io.pixelsdb.pixels.core.vector.LongDecimalColumnVector cannot be cast to class io.pixelsdb.pixels.core.vector.DecimalColumnVector (io.pixelsdb.pixels.core.vector.LongDecimalColumnVector and io.pixelsdb.pixels.core.vector.DecimalColumnVector are in unnamed module of loader lambdainternal.CustomerClassLoader @76fb509a)
	at io.pixelsdb.pixels.executor.aggregation.function.DecimalSum.input(DecimalSum.java:53) ~[task/:?]
	at io.pixelsdb.pixels.executor.aggregation.AggrTuple.initFunctions(AggrTuple.java:71) ~[task/:?]
	at io.pixelsdb.pixels.executor.aggregation.Aggregator.aggregate(Aggregator.java:186) ~[task/:?]
	at io.pixelsdb.pixels.worker.common.BaseAggregationWorker.aggregate(BaseAggregationWorker.java:284) ~[task/:?]
	at io.pixelsdb.pixels.worker.common.BaseAggregationWorker.lambda$process$0(BaseAggregationWorker.java:173) ~[task/:?]
	... 3 more

```

--- a/pixels-core/src/main/java/io/pixelsdb/pixels/core/utils/Integer128.java
+++ b/pixels-core/src/main/java/io/pixelsdb/pixels/core/utils/Integer128.java
@@ -73,22 +73,26 @@ public class Integer128 implements Comparable<Integer128>
      */
     public static Integer128 fromBigEndian(byte[] bytes)
     {
-        if (bytes.length >= 16) {
+        if (bytes.length >= 16)
+        {
             int offset = bytes.length - Long.BYTES;
             long low = fromBigEndianBytes(bytes, offset);
 
             offset -= Long.BYTES;
             long high = fromBigEndianBytes(bytes, offset);
 
-            for (int i = 0; i < offset; i++) {
-                if (bytes[i] != (high >> 63)) {
+            for (int i = 0; i < offset; i++)
+            {
+                if (bytes[i] != (high >> 63))
+                {
                     throw new ArithmeticException("Overflow");
                 }
             }
 
             return Integer128.valueOf(high, low);
         }
-        else if (bytes.length > 8) {
+        else if (bytes.length > 8)
+        {
             // read the last 8 bytes into low
             int offset = bytes.length - Long.BYTES;
             long low = fromBigEndianBytes(bytes, offset);
@@ -102,17 +106,20 @@ public class Integer128 implements Comparable<Integer128>
 
             return Integer128.valueOf(high, low);
         }
-        else if (bytes.length == 8) {
+        else if (bytes.length == 8)
+        {
             long low = fromBigEndianBytes(bytes, 0);
             long high = (low >> 63);
 
             return Integer128.valueOf(high, low);
         }
-        else {
+        else
+        {
             long high = (bytes[0] >> 7);
             long low = high;
-            for (int i = 0; i < bytes.length; i++) {
-                low = (low << 8) | (bytes[i] & 0xFF);
+            for (byte aByte : bytes)
+            {
+                low = (low << 8) | (aByte & 0xFF);
             }
 
             return Integer128.valueOf(high, low);
@@ -121,7 +128,8 @@ public class Integer128 implements Comparable<Integer128>
 
     public static Integer128 valueOf(long[] value)
     {
-        if (value.length != 2) {
+        if (value.length != 2)
+        {
             throw new IllegalArgumentException("Expected long[2]");
         }
 
@@ -144,7 +152,8 @@ public class Integer128 implements Comparable<Integer128>
     {
         long low = value.longValue();
         long high;
-        try {
+        try
+        {
             high = value.shiftRight(64).longValueExact();
         }
         catch (ArithmeticException e) {
@@ -161,10 +170,10 @@ public class Integer128 implements Comparable<Integer128>
 
     public void add(long high, long low)
     {
-        // TODO: optimize.
         BigInteger res = this.toBigInteger().add(new BigInteger(toBigEndianBytes(high, low)));
         this.low = res.longValue();
-        try {
+        try
+        {
             this.high = res.shiftRight(64).longValueExact();
         }
         catch (ArithmeticException e) {
@@ -175,10 +184,12 @@ public class Integer128 implements Comparable<Integer128>
     @Override
     public boolean equals(Object o)
     {
-        if (this == o) {
+        if (this == o)
+        {
             return true;
         }
-        if (o == null || getClass() != o.getClass()) {
+        if (o == null || getClass() != o.getClass())
+        {
             return false;
         }
         Integer128 that = (Integer128) o;
@@ -224,7 +235,8 @@ public class Integer128 implements Comparable<Integer128>
 
     public long toLongExact()
     {
-        if (high != (low >> 63)) {
+        if (high != (low >> 63))
+        {
             throw new ArithmeticException("Overflow");
         }
 
@@ -239,7 +251,8 @@ public class Integer128 implements Comparable<Integer128>
     public static int compare(long leftHigh, long leftLow, long rightHigh, long rightLow)
     {
         int comparison = Long.compare(leftHigh, rightHigh);
-        if (comparison == 0) {
+        if (comparison == 0)
+        {
             comparison = Long.compareUnsigned(leftLow, rightLow);
         }
 
@@ -266,7 +279,8 @@ public class Integer128 implements Comparable<Integer128>
 
     public static void toBigEndianBytes(long value, byte[] bytes, int offset)
     {
-        for(int i = 7; i >= 0; --i) {
+        for(int i = 7; i >= 0; --i)
+        {
             bytes[offset+i] = (byte)((int)(value & 255L));
             value >>= 8;
         }

--- a/pixels-executor/src/main/java/io/pixelsdb/pixels/executor/aggregation/function/DecimalSum.java
+++ b/pixels-executor/src/main/java/io/pixelsdb/pixels/executor/aggregation/function/DecimalSum.java
@@ -23,6 +23,10 @@ import io.pixelsdb.pixels.core.TypeDescription;
 import io.pixelsdb.pixels.core.utils.Integer128;
 import io.pixelsdb.pixels.core.vector.ColumnVector;
 import io.pixelsdb.pixels.core.vector.DecimalColumnVector;
+import io.pixelsdb.pixels.core.vector.LongDecimalColumnVector;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static io.pixelsdb.pixels.core.TypeDescription.MAX_SHORT_DECIMAL_SCALE;
 
 /**
  * @author hank
@@ -30,19 +34,32 @@ import io.pixelsdb.pixels.core.vector.DecimalColumnVector;
  */
 public class DecimalSum extends SingleColumnFunction
 {
+    private final TypeDescription inputType;
     private final TypeDescription outputType;
+    private final boolean inputIsLong;
+    private final boolean outputIsLong;
     private Integer128 longValue;
     private long shortValue = 0;
-    private boolean isLong = false;
 
-    protected DecimalSum(TypeDescription outputType)
+    protected DecimalSum(TypeDescription inputType, TypeDescription outputType)
     {
+        this.inputType = inputType;
         this.outputType = outputType;
-        if (outputType.getPrecision() > 18)
+        this.inputIsLong = inputType.getPrecision() > MAX_SHORT_DECIMAL_SCALE;
+
+        if (outputType.getPrecision() > MAX_SHORT_DECIMAL_SCALE)
         {
-            this.isLong = true;
+            this.outputIsLong = true;
             this.longValue = new Integer128(0L, 0L);
         }
+        else
+        {
+            this.outputIsLong = false;
+        }
+
+        // Issue #647: check input and output data type.
+        checkArgument(!this.inputIsLong || this.outputIsLong,
+                "the output type must be long decimal if the input type is long decimal");
     }
 
     @Override
@@ -50,14 +67,23 @@ public class DecimalSum extends SingleColumnFunction
     {
         if (inputVector.noNulls || !inputVector.isNull[rowId])
         {
-            DecimalColumnVector decimalColumnVector = (DecimalColumnVector) inputVector;
-            if (isLong)
+            if (inputIsLong)
             {
-                longValue.add(decimalColumnVector.vector[rowId*2], decimalColumnVector.vector[rowId*2+1]);
+                LongDecimalColumnVector decimalColumnVector = (LongDecimalColumnVector) inputVector;
+                // Issue #647: if the input type is long decimal, the output type must be long decimal too.
+                longValue.add(decimalColumnVector.vector[rowId * 2], decimalColumnVector.vector[rowId * 2 + 1]);
             }
             else
             {
-                shortValue += decimalColumnVector.vector[rowId];
+                DecimalColumnVector decimalColumnVector = (DecimalColumnVector) inputVector;
+                // Issue #647: if the input type is short decimal, the output could be short or long decimal.
+                if (outputIsLong)
+                {
+                    longValue.add(0L, decimalColumnVector.vector[rowId]);
+                } else
+                {
+                    shortValue += decimalColumnVector.vector[rowId];
+                }
             }
         }
     }
@@ -65,7 +91,7 @@ public class DecimalSum extends SingleColumnFunction
     @Override
     public void output(ColumnVector outputVector)
     {
-        if (isLong)
+        if (outputIsLong)
         {
             outputVector.add(longValue);
         }
@@ -78,6 +104,6 @@ public class DecimalSum extends SingleColumnFunction
     @Override
     public Function buildCopy()
     {
-        return new DecimalSum(this.outputType);
+        return new DecimalSum(this.inputType, this.outputType);
     }
 }

--- a/pixels-executor/src/main/java/io/pixelsdb/pixels/executor/aggregation/function/FunctionFactory.java
+++ b/pixels-executor/src/main/java/io/pixelsdb/pixels/executor/aggregation/function/FunctionFactory.java
@@ -38,7 +38,7 @@ public class FunctionFactory
     private FunctionFactory() { }
 
     public Function createFunction(FunctionType functionType,
-                                   TypeDescription inputType,TypeDescription outputType)
+                                   TypeDescription inputType, TypeDescription outputType)
     {
         switch (functionType)
         {
@@ -62,7 +62,7 @@ public class FunctionFactory
             case LONG:
                 return new BigintSum();
             case DECIMAL:
-                return new DecimalSum(outputType);
+                return new DecimalSum(inputType, outputType);
             default:
                 // TODO: support more types.
                 throw new UnsupportedOperationException(

